src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -677,5 +723,49 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
                 throw new NotSupportedException();
             }
         }
+
+        private class DuplexPipeConverter : JsonConverter<IDuplexPipe>
+        {
+            private readonly JsonMessageFormatter jsonMessageFormatter;
+
+            public DuplexPipeConverter(JsonMessageFormatter jsonMessageFormatter)
+            {
+                this.jsonMessageFormatter = jsonMessageFormatter ?? throw new ArgumentNullException(nameof(jsonMessageFormatter));
+            }
+
+            public override IDuplexPipe ReadJson(JsonReader reader, Type objectType, IDuplexPipe existingValue, bool hasExistingValue, JsonSerializer serializer)
+            {
+                int? tokenId = JToken.Load(reader).Value<int?>();
+                return this.jsonMessageFormatter.duplexPipeTracker.GetPipe(tokenId);
+            }
+
+            public override void WriteJson(JsonWriter writer, IDuplexPipe value, JsonSerializer serializer)
+            {
+                var token = this.jsonMessageFormatter.duplexPipeTracker.GetToken(value);
+                writer.WriteValue(token);
+            }
+        }
+
+        private class StreamConverter : JsonConverter<Stream>
+        {
+            private readonly JsonMessageFormatter jsonMessageFormatter;
+
+            public StreamConverter(JsonMessageFormatter jsonMessageFormatter)
+            {
+                this.jsonMessageFormatter = jsonMessageFormatter ?? throw new ArgumentNullException(nameof(jsonMessageFormatter));
+            }
+
+            public override Stream ReadJson(JsonReader reader, Type objectType, Stream existingValue, bool hasExistingValue, JsonSerializer serializer)
+            {
+                int? tokenId = JToken.Load(reader).Value<int?>();
+                return this.jsonMessageFormatter.duplexPipeTracker.GetPipe(tokenId)?.AsStream();
+            }
+
+            public override void WriteJson(JsonWriter writer, Stream value, JsonSerializer serializer)
+            {
+                var token = this.jsonMessageFormatter.duplexPipeTracker.GetToken(value?.UsePipe());
```

Comment:
Using a stream as a pipe loses its random access capability, so we can either not allow such streams to be passed through, or document this behavior.

How about one-directional streams, that can only read or write? Looks like write-only streams are also not supported, because `UsePipeReader` in `PipeExtensions.cs` in `Nerdbank.Streams` would throw.

Reply:
> Using a stream as a pipe loses its random access capability

What do you mean? `Stream.UsePipe()` returns an `IDuplexPipe` which can be read from and written to just like a duplex `Stream`. The only lossy conversion is from `IDuplexPipe` to `Stream`, since only `IDuplexPipe` lets you 'complete' one direction of communication and a `Stream` doesn't offer such an API.
And I *do* document that in the .md file I add in this PR.

> How about one-directional streams, that can only read or write? 

I support that too. Check out the unit tests I add. I specifically test with both read-only and write-only streams *and* "duplex" pipes. It all works.
I [updated Nerdbank.Streams' `UsePipe()` extension method](https://github.com/AArnott/Nerdbank.Streams/pull/115) to support one-directional streams last week to support this.

Reply:
>> Using a stream as a pipe loses its random access capability

> What do you mean?

Suppose some client passes a stream, and then wants to set `Position` on it. Will this screw up the service? Can the service use `Position` on the stream?

>> How about one-directional streams, that can only read or write?

> I support that too.
Cool. My local copy of `Nerdbank.Streams` doesn't have it, that's why I was asking.

Reply:
> Suppose some client passes a stream, and then wants to set Position on it. 

Network streams are not seekable. `Stream.CanSeek == false`. So you can't set `Position` on it -- it would throw `NotSupportedException`. Seek only makes sense on a stream representing a persistent thing like a file. But these streams are network streams -- they represent communication channels, and thus cannot seek, but you can read and write to them concurrently, which does *not* make sense for a file.

IMO it's confusing how in .NET a Stream can represent two totally different use cases like this. I think that is somewhat alleviated by the introduction of `IDuplexPipe` which is a type dedicated to just one of these (and does a better job).